lib/iris/analysis/__init__.py

```diff
@@ -1419,36 +1419,42 @@ def interp_order(length):
 """
 
 
-def _build_dask_mdtol_function(dask_nanfunction):
+def _build_dask_mdtol_function(dask_stats_function):
     """
     Make a wrapped dask statistic function that supports the 'mdtol' keyword.
 
-    'dask_nanfunction' must be a statistical function that treats NaNs as
-    missing data, and which has the basic call signature
-    "dask_nanfunction(data, axis, **kwargs)".
+    'dask_function' must be a dask statistical function, compatible with the
+    call signature : "dask_stats_function(data, axis, **kwargs)".
+    It must be masked-data tolerant, i.e. it ignores masked input points and
+    performs a calculation on only the unmasked points.
+    For example, mean([1, --, 2]) = (1 + 2) / 2 = 1.5.
 
     The returned value is a new function operating on dask arrays.
-    It has the call signature "stat(data, axis=-1, mdtol=None, *kwargs)".
+    It has the call signature "stat(data, axis=-1, mdtol=None, **kwargs)".
 
     """
-    @wraps(dask_nanfunction)
+    @wraps(dask_stats_function)
     def inner_stat(array, axis=-1, mdtol=None, **kwargs):
-        dask_result = dask_nanfunction(array, axis=axis, **kwargs)
+        # Call the statistic to get the basic result (missing-data tolerant).
+        dask_result = dask_stats_function(array, axis=axis, **kwargs)
         if mdtol is None:
             result = dask_result
         else:
-            point_counts = np.sum(np.ones(array.shape), axis=axis)
-            point_mask_counts = da.sum(da.isnan(array), axis=axis)
-            masked_point_fractions = (point_mask_counts + 0.0) / point_counts
-            # Note: the +0.0 forces a floating-point divide.
+            # Build a lazy computation to compare the fraction of missing
+            # input points at each output point to the 'mdtol' threshold.
+            point_counts = da.sum(da.ones(array.shape, chunks=array.chunks),
```

Comment:
Does this need to be lazy?

Reply:
This shouldn't be filling the array at all. This is the only real blocker to this being merged IMO.

Reply:
How about you get the consistency without the cost of the ones...

```
>>> shape = [1201230123, 123123, 123123, 1231, 3123123]
>>> a = da.empty(shape, chunks=shape)
>>> 
>>> a.shape
(1201230123, 123123, 123123, 1231, 3123123)
>>> r = da.sum(a, axis=[1, 2])
>>> r.shape
(1201230123, 1231, 3123123)
```